File: heartbeat/src/service_ping.rs
use crate::SERVICE_ID; // To access the SERVICE_ID from main.rs
use hostname;
use libsql::{params, Builder};
use std::time::SystemTime;
use tokio::time::{sleep, Duration};
use tracing::{info, warn};

const HEALTH_PING_INTERVAL_SECS: u64 = 30;

pub async fn run_service_ping_task(db_url: String, auth_token: String) {
    let host_name = hostname::get()
        .map(|s| {
            s.into_string()
                .unwrap_or_else(|_| "invalid_hostname".to_string())
        })
        .unwrap_or_else(|_| "unknown_hostname".to_string());

    info!(
        "ServicePing task started for service_id '{}'. Will ping DB at {} (auth token used: {}) every {}s. Hostname: {}",
        SERVICE_ID, db_url, !auth_token.is_empty(), HEALTH_PING_INTERVAL_SECS, host_name
    );

    loop {
        sleep(Duration::from_secs(HEALTH_PING_INTERVAL_SECS)).await;

        let db_result = async {
            let db = Builder::new_remote(db_url.clone(), auth_token.clone())
                .build()
                .await?;
            let conn = db.connect()?;

            let app_timestamp = SystemTime::now()
                .duration_since(SystemTime::UNIX_EPOCH)
                .map(|d| d.as_secs())
                .unwrap_or(0);

            info!(
                "Sending ServicePing for service_id '{}', hostname '{}'",
                SERVICE_ID, host_name
            );
            conn.execute(
                "INSERT INTO ServicePing (service_id, hostname, app_timestamp) VALUES (?1, ?2, ?3)",
                params![SERVICE_ID, host_name.clone(), app_timestamp as i64],
            )
            .await?;

            Ok::<(), Box<dyn std::error::Error + Send + Sync>>(())
        }
        .await;

        if let Err(e) = db_result {
            warn!("Failed to send ServicePing: {}", e);
        }
    }
}

File: heartbeat/service_ping.go
package main

import (
	"context"
	"database/sql"
	"log/slog"
	"os"
	"time"

	_ "github.com/tursodatabase/libsql-client-go/libsql"
)

const healthPingInterval = 30 * time.Second

func runServicePingTask(dbURL, authToken string) {
	host, err := os.Hostname()
	if err != nil || host == "" {
		host = "unknown_hostname"
	}

	slog.Info("ServicePing task started",
		"service_id", SERVICE_ID,
		"db", dbURL,
		"auth_token_provided", authToken != "",
		"interval_secs", int(healthPingInterval.Seconds()),
		"hostname", host,
	)

	// Each tick: open a fresh connection (keeps logic closest to the Rust task),
	// insert a row, then close. You can also reuse a shared *sql.DB if preferred.
	ticker := time.NewTicker(healthPingInterval)
	defer ticker.Stop()

	for {
		<-ticker.C

		remoteURL, err := buildRemoteURL(dbURL, authToken)
		if err != nil {
			slog.Warn("ServicePing: invalid DB URL", "error", err)
			continue
		}

		db, err := sql.Open("libsql", remoteURL)
		if err != nil {
			slog.Warn("ServicePing: failed to open DB", "error", err)
			continue
		}

		appTimestamp := time.Now().Unix()

		slog.Info("Sending ServicePing", "service_id", SERVICE_ID, "hostname", host)

		execCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		_, execErr := db.ExecContext(execCtx,
			`INSERT INTO ServicePing (service_id, hostname, app_timestamp)
			 VALUES (?, ?, ?)`,
			SERVICE_ID, host, appTimestamp,
		)
		cancel()

		if execErr != nil {
			slog.Warn("Failed to send ServicePing", "error", execErr)
		}

		closeErr := db.Close()
		if closeErr != nil {
			slog.Warn("ServicePing: failed closing DB", "error", closeErr)
		}
	}
}
